src/librbd/Group.cc

```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+				    const char *group_name,
+				    const char *snap_name)
+{
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+
+  std::vector<group_snap_spec_t> snaps;
+  int r = group_snap_list(group_ioctx, group_name, snaps);
+  if (r < 0) {
+    lderr(cct) << "failed to list existing snapshots while checking name duplicates: "
+	       << cpp_strerror(r)
+	       << dendl;
+    return r;
+  }
+  for (auto i: snaps) {
+    if (i.name == string(snap_name)) {
+      lderr(cct) << "snapshot with this name already exists: "
+		 << cpp_strerror(r)
+		 << dendl;
+      return -EEXIST;
+    }
+  }
+  return 0;
+}
+
+std::string calc_ind_image_snap_name(uint64_t pool_id,
+				     std::string group_id,
+				     uint64_t snap_seq)
+{
+  std::stringstream ind_snap_name_stream;
+  ind_snap_name_stream << std::setw(16) << std::setfill('0') << std::hex <<
+			  pool_id <<
+			  "_" << group_id <<
+			  "_" <<
+			  std::setw(16) << std::setfill('0') << std::hex <<
+			  snap_seq;
+  return ind_snap_name_stream.str();
+}
+
+int group_snap_create(librados::IoCtx& group_ioctx,
+		      const char *group_name, const char *snap_name)
+{
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  librados::Rados rados(group_ioctx);
+
+  int r = group_snap_name_check_duplicate(group_ioctx, group_name, snap_name);
+  if (r < 0) {
+    return r;
+  }
+
+  string group_id;
+  uint64_t snap_seq = 0;
+  cls::rbd::GroupSnapshot gs;
+  vector<cls::rbd::ImageSnapshotRef> image_snaps;
+  std::string ind_snap_name;
+
+  r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+			     group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading consistency group id object: "
+	       << cpp_strerror(r)
+	       << dendl;
+    return r;
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  std::vector<group_image_status_t> images;
+  r = librbd::group_image_list(group_ioctx, group_name, images);
+  if (r < 0) {
+    return r;
+  }
+  int n = images.size();
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+  std::vector<C_SaferCond*> on_finishes;
+  for (auto i: images) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
+
+    r = rados.ioctx_create2(i.pool, *image_io_ctx);
+    if (r < 0) {
+      ldout(cct, 1) << "Failed to create io context for image" << dendl;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx(i.name.c_str(), "", "",
+					       *image_io_ctx, false);
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    image_ctx->state->open(on_finish);
+
+    ictxs.push_back(image_ctx);
+    on_finishes.push_back(on_finish);
+  }
+  int ret_code = 0;
+  for (int i = 0; i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0) {
+      delete ictxs[i];
+      ictxs[i] = nullptr;
+      ret_code = r;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+  for (auto i: ictxs) {
+    i->exclusive_lock->block_requests(-EBUSY);
```

Comment:
Need to hold ```owner_lock``` when touching ```ImageCtx::exclusive_lock``` -- and should verify that it isn't a nullptr before dereferencing. Same comment applies below.